python/src/ops.cpp

```diff
@@ -2280,6 +2280,38 @@ void init_ops(py::module_& m) {
         Returns:
             list(array): A list of split arrays.
       )pbdoc");
+  m.def(
+      "fast_inference_sdpa",
+      &fast_inference_sdpa,
+      "q"_a,
+      "k"_a,
+      "v"_a,
+      "scale"_a,
+      "mask"_a = none,
+      py::kw_only(),
+      "stream"_a = none,
+      R"pbdoc(
+          fast_inference_sdpa(q: array, k: array, v: array,  /, *, stream: Union[None, Stream, Device] = None) -> array
+
+          A parallelized implementation of multi-head attention: O = softmax(Q @ K.T, dim=-1) @ V,
+          with a split along the KV sequence length.  Supports MHA, GQA, and MQA.
+
+          Perform the (possibly batched) matrix multiplication of two arrays. This function is an inference-only
+          kernel optimized specifically for KV-cached transformer decoder inference (query sequence length = 1) and
+          large KV-cached sequences.  It handles prompt encoding via MLX primitives, optimized metal kernel
+          is for decoding only.
```

Comment:
The first sentence is out of place? I think if we do the fallback + custom primitive then we shouldn't need to caveat the application of this function for inference only. So you could probably remove this entire paragraph.

Reply:
yes, can remove that.

the GPU shaders would remain supporting inference only though? unless there is a way to route to the MLX primitives sequence, as of now only the prompt encoding or CPU impl would go through mlx primitives?

Reply:
Yea that's right. So the way it would work is in the `sdpm` op itself you would route to the fast primitive if the conditions made sense (e.g. query was a single example, supported data type, etc) otherwise you call the fallback. Here's [an example](https://github.com/ml-explore/mlx/blob/main/mlx/fast.cpp#L111-L119).

It might be nice add a comment about when the actual fast op will be used, but I would put it in terms of constraints on the inputs (rather than "inference only"), we can update it if the conditions change.

Reply:
ok nice!  yes that's a great example, perfect.